tools/mkcomposefs.c

```diff
@@ -760,46 +760,25 @@ static int mkdir_parents(const char *pathname, int mode)
 	return 0;
 }
 
-static int write_to_fd(int fd, const char *content, ssize_t len)
+static int copy_file_data(int sfd, int dfd)
 {
-	ssize_t res;
-
-	while (len > 0) {
-		res = write(fd, content, len);
-		if (res < 0 && errno == EINTR)
-			continue;
-		if (res <= 0) {
-			if (res == 0) /* Unexpected short write, should not happen when writing to a file */
-				errno = ENOSPC;
-			return -1;
-		}
-		len -= res;
-		content += res;
-	}
+	struct stat stat;
 
-	return 0;
-}
+	if (fstat(sfd, &stat) == -1)
+		return -1;
 
-#define BUFSIZE 8192
-static int copy_file_data(int sfd, int dfd)
-{
-	char buffer[BUFSIZE];
-	ssize_t bytes_read;
-
-	while (true) {
-		bytes_read = read(sfd, buffer, BUFSIZE);
-		if (bytes_read == -1) {
-			if (errno == EINTR)
-				continue;
-			return -1;
-		}
+	off_t len, ret;
+	len = stat.st_size;
 
-		if (bytes_read == 0)
+	do {
+		ret = copy_file_range(sfd, NULL, dfd, NULL, len, 0);
+		if (ret == -1)
```

Comment:
I wonder also, goes glibc have a fallback for older kernels for this?

Reply:
Yeah, seems like it.

Reply:
EINTR is not mentioned as a possible error for copy_file_range. Should that be handled?

Reply:
I'm not sure if the syscall is guaranteed to never get EINTR (although I doubt it), but the glibc fallback implementation clearly doesn't handle it. So, yes, we need to handle EINTR.

Reply:
Fixed. Thank you

Reply:
I think we still need a fallback to the older code, since `copy_file_range` doesn't seem to work across different file systems:

```
$ ./tools/mkcomposefs --digest-store=/tmp/store  tests /dev/null
mkcomposefs: cannot fill store: Invalid cross-device link
```

Reply:
Do you suggest checking for error code ```EXDEV``` and fallback to previous implementation of read and write or totally replacing copy_file_range with read/write? copy_file_range is already a fallback for soft link attempt. Overall performance gain in my case was ~1% with this api change(i didn't run a lot of bench marks just for this).

Reply:
> copy_file_range is already a fallback for soft link attempt.

that still fails with different file systems; and I think it is good to keep the new code as it is faster.

I don't think we can add the fallback for any error as the fds could be already used, but `EXDEV` looks safe.

Reply:
It's worth using existing codebases as a reference for which errno values to use as fallback:

- Rust: https://github.com/rust-lang/rust/blob/0e5f5207881066973486e6a480fa46cfa22947e9/library/std/src/sys/pal/unix/kernel_copy.rs#L622
- libglnx: https://gitlab.gnome.org/GNOME/libglnx/-/blob/202b294e6079e23242e65e0426f8639841d1210b/glnx-fdio.c#L846 (which was long ago forked from systemd):
- systemd: https://github.com/systemd/systemd/blob/e71b40fd0026c0884ca26eb4f0a9fbe4d9285cfa/src/shared/copy.c#L338

Reply:
Thank you for the links. Reading through https://lwn.net/Articles/846403/ i think it is safer to rollback `copy_file_range` to older `read/write` or perhaps address this as a separate PR as this does not have a direct dependency to the original multi-threading change. What do you say? I am very new to linux and i was looking for better apis to copy a file without user mode buffer.